Name: Muley, Tushar

Title: Exercise_9-3_Week_9

Date: May 23, 2021


# 1. Neural Network Classifier with Scikit

Using the multi-label classifier dataset from earlier exercises (categorized-comments.jsonl in the reddit folder), fit a neural network classifier using scikit-learn. Use the code found in chapter 12 of the Applied Text Analysis with Python book as a guideline. Report the accuracy, precision, recall, F1-score, and confusion matrix.

In [3]:
#Import libraries
import numpy as np # For numerical fast numerical calculations
import matplotlib.pyplot as plt # For making plots
import pandas as pd # Deals with data
import seaborn as sns # Makes beautiful plots
from sklearn.preprocessing import StandardScaler # Testing sklearn
import tensorflow # Imports tensorflow
import keras # Imports keras

In [4]:
#Read the file
jfile = "categorized-comments.jsonl" #File location
file = pd.read_json(jfile,lines=True) #Read command
file.head() #Check the datarame

cat                                                txt
0  sports  Barely better than Gabbert? He was significant...
1  sports  Fuck the ducks and the Angels! But welcome to ...
2  sports  Should have drafted more WRs.\n\n- Matt Millen...
3  sports            [Done](https://i.imgur.com/2YZ90pm.jpg)
4  sports                                      No!! NOO!!!!!

Using the multi-label classifier dataset from earlier exercises (categorized-comments.jsonl in the reddit folder),

In [3]:
#Size of the dataframe
file.shape

(606476, 2)

In [5]:
#Lowercase the data
file["txt"] = file["txt"].str.lower()
#View the top 5 rows

#Remove punctuation
file["txt"] = file["txt"].str.replace("[^\w\s]", "")
#View the top 5 rows
file.head()

cat                                                txt
0  sports  barely better than gabbert he was significantl...
1  sports  fuck the ducks and the angels but welcome to a...
2  sports  should have drafted more wrs\n\n matt millen p...
3  sports                       donehttpsiimgurcom2yz90pmjpg
4  sports                                             no noo

In [6]:
#Download stopwords and punkt
import nltk
nltk.download('stopwords')

import nltk
nltk.download('punkt')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Tushar\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Tushar\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [7]:
#Import NLKT, stopwords and tokenize
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 

#Set the stop_words
stop_words = stopwords.words("english") 

In [8]:
#Tokenize the text
file["tokenized_sents"] = file.apply(lambda row: word_tokenize(row["txt"]), axis=1)

#Preview the file
file.head()

cat                                                txt  \
0  sports  barely better than gabbert he was significantl...   
1  sports  fuck the ducks and the angels but welcome to a...   
2  sports  should have drafted more wrs\n\n matt millen p...   
3  sports                       donehttpsiimgurcom2yz90pmjpg   
4  sports                                             no noo   

                                     tokenized_sents  
0  [barely, better, than, gabbert, he, was, signi...  
1  [fuck, the, ducks, and, the, angels, but, welc...  
2  [should, have, drafted, more, wrs, matt, mille...  
3                     [donehttpsiimgurcom2yz90pmjpg]  
4                                          [no, noo]

In [8]:
#Remove stop words
file["removed_stop_words"]=[word for word in file["tokenized_sents"] if word not in stop_words]

#Preview words
file.head()

cat                                                txt  \
0  sports  barely better than gabbert he was significantl...   
1  sports  fuck the ducks and the angels but welcome to a...   
2  sports  should have drafted more wrs\n\n matt millen p...   
3  sports                       donehttpsiimgurcom2yz90pmjpg   
4  sports                                             no noo   

                                     tokenized_sents  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                  removed_stop_words  
0  [barely, better, than, gabbert, he, was, signi...  
1  [fuck, the, ducks, and, the, angels, but, welc...  
2  [should, have, drafted, more, wrs, matt, mille...  
3                     [donehttpsiimgurcom2yz90pmjpg]  
4                                          [no, noo]

In [9]:
#Import PorterStemmer
from nltk.stem.porter import PorterStemmer

#Create stemmer
porter = PorterStemmer()

In [10]:
#Run stemmer
file["stemmer"]= file["removed_stop_words"].apply(lambda x: [porter.stem(word) for word in x])
#View the file
file.head()

cat                                                txt  \
0  sports  barely better than gabbert he was significantl...   
1  sports  fuck the ducks and the angels but welcome to a...   
2  sports  should have drafted more wrs\n\n matt millen p...   
3  sports                       donehttpsiimgurcom2yz90pmjpg   
4  sports                                             no noo   

                                     tokenized_sents  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                  removed_stop_words  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                             stemmer  
0  [bare, better, than, gabbert, he, wa, signific...  
1  [fuck, the, duck, and, the, angel, but, welcom...  
2  [should, have, draft, more, wr, matt, millen, ...  
3                     [donehttpsiimgurcom2yz90pmjpg]  
4                                          [no, noo]

In [11]:
#Check the size of the data
file.shape

(606476, 5)

In [12]:
#Copy of file check point
df_file_copy = file.copy(deep=True)

#Only use if needed to get orignal copy
#file = df_file_copy.copy(deep=True)

#Preview
file.head()

cat                                                txt  \
0  sports  barely better than gabbert he was significantl...   
1  sports  fuck the ducks and the angels but welcome to a...   
2  sports  should have drafted more wrs\n\n matt millen p...   
3  sports                       donehttpsiimgurcom2yz90pmjpg   
4  sports                                             no noo   

                                     tokenized_sents  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                  removed_stop_words  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                             stemmer  
0  [bare, better, than, gabbert, he, wa, signific...  
1  [fuck, the, duck, and, the, angel, but, welcom...  
2  [should, have, draft, more, wr, matt, millen, ...  
3                     [donehttpsiimgurcom2yz90pmjpg]  
4                                          [no, noo]

In [13]:
#put the text back together string it
file["txt_final"]= file["stemmer"].apply(lambda text:' '.join(text))
#view the pre-processed text
file.head()

cat                                                txt  \
0  sports  barely better than gabbert he was significantl...   
1  sports  fuck the ducks and the angels but welcome to a...   
2  sports  should have drafted more wrs\n\n matt millen p...   
3  sports                       donehttpsiimgurcom2yz90pmjpg   
4  sports                                             no noo   

                                     tokenized_sents  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                  removed_stop_words  \
0  [barely, better, than, gabbert, he, was, signi...   
1  [fuck, the, ducks, and, the, angels, but, welc...   
2  [should, have, drafted, more, wrs, matt, mille...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                             stemmer  \
0  [bare, better, than, gabbert, he, wa, signific...   
1  [fuck, the, duck, and, the, angel, but, welcom...   
2  [should, have, draft, more, wr, matt, millen, ...   
3                     [donehttpsiimgurcom2yz90pmjpg]   
4                                          [no, noo]   

                                           txt_final  
0  bare better than gabbert he wa significantli b...  
1  fuck the duck and the angel but welcom to all ...  
2      should have draft more wr matt millen probabl  
3                       donehttpsiimgurcom2yz90pmjpg  
4                                             no noo

In [14]:
#Copy data check point
df_file_copy2 = file.copy(deep=True)

In [15]:
#Drop un-need columns

file=file.drop(["txt", "tokenized_sents", "removed_stop_words", "stemmer"], axis=1)
file.head()

cat                                          txt_final
0  sports  bare better than gabbert he wa significantli b...
1  sports  fuck the duck and the angel but welcom to all ...
2  sports      should have draft more wr matt millen probabl
3  sports                       donehttpsiimgurcom2yz90pmjpg
4  sports                                             no noo

In [36]:
#Check size of file
file.shape

(606476, 2)

In [37]:
#Sample to redue the size of the data set
file_small=file.sample(n = 60000)

In [38]:
#Check size
file_small.shape

(60000, 2)

In [40]:
#Preview the data set
file_small.head()

cat                                          txt_final
408464  video_games  ha it been announc if sonic mania is get physi...
545789  video_games  that half true but pro team with max money on ...
157080  video_games                        whatev float your boat mate
138498  video_games       cheeki light up the backsid at the end there
590515  video_games   httpplaystvvideo585e42248d71bde305360fromus same

In [16]:
#Bring in CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

In [41]:
#Create the bag of words features Features = X
count=CountVectorizer()
X=count.fit_transform(file_small["txt_final"])
X.shape

(60000, 39535)

In [96]:
X

<60000x39535 sparse matrix of type '<class 'numpy.int64'>'
	with 1221428 stored elements in Compressed Sparse Row format>

In [97]:
#All my targets Target = y
y=file_small["cat"]
y.shape

(60000,)

In [98]:
y

408464               video_games
545789               video_games
157080               video_games
138498               video_games
590515               video_games
                   ...          
437435               video_games
22592     science_and_technology
526688               video_games
589654               video_games
406959               video_games
Name: cat, Length: 60000, dtype: object

In [43]:
#Import train_test_split
from sklearn.model_selection import train_test_split

#Split the data into train and text dataset
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size = 0.20, random_state = 1)

In [44]:
#Preview the train test dataset
print('X Training set:',X_train.shape)

print ('Y Training set:',y_train.shape)

print ('X Test:',X_test.shape)

print ('Y Test:',y_test.shape)

X Training set: (48000, 39535)
Y Training set: (48000,)
X Test: (12000, 39535)
Y Test: (12000,)


In [45]:
#Import standard Scaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler(with_mean=False)

#Fit only the training data
scaler.fit(X_train)

StandardScaler(with_mean=False)

In [46]:
#Apply the transformations to data:
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [47]:
#Import classeifier
from sklearn.neural_network import MLPClassifier

In [28]:
#Instance of the model.
mlp = MLPClassifier(hidden_layer_sizes=(5,2))
#Fit the training data to our model
mlp.fit(X_train,y_train)

C:\Users\Tushar\anaconda3\envs\PythonCPU\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:587: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


MLPClassifier(hidden_layer_sizes=(5, 2))

In [29]:
#Model is trained now test the model.
predictions = mlp.predict(X_test)

In [30]:
#Import classification and confusion matrix
from sklearn.metrics import classification_report,confusion_matrix

print(confusion_matrix(y_test,predictions))
 
print(classification_report(y_test,predictions))

#print(accuracy_score(y_test,predictions))

[[  3455    129   2734]
 [   130  23473  12899]
 [  1084   5158 102557]]
                        precision    recall  f1-score   support

science_and_technology       0.74      0.55      0.63      6318
                sports       0.82      0.64      0.72     36502
           video_games       0.87      0.94      0.90    108799

              accuracy                           0.85    151619
             macro avg       0.81      0.71      0.75    151619
          weighted avg       0.85      0.85      0.85    151619



In [48]:
#Instance of the model.
mlp = MLPClassifier(hidden_layer_sizes=(30,30,30),verbose=False)
#Fit the training data to our model
mlp.fit(X_train,y_train)

MLPClassifier(hidden_layer_sizes=(30, 30, 30))

In [49]:
#Model is trained now test the model.
predictions = mlp.predict(X_test)

In [50]:
#Import classification and confusion matrix
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

print(confusion_matrix(y_test,predictions))
 
print(classification_report(y_test,predictions))
 
print(accuracy_score(y_test,predictions))

[[ 223   42  239]
 [  43 1781 1078]
 [ 142  873 7579]]
                        precision    recall  f1-score   support

science_and_technology       0.55      0.44      0.49       504
                sports       0.66      0.61      0.64      2902
           video_games       0.85      0.88      0.87      8594

              accuracy                           0.80     12000
             macro avg       0.69      0.65      0.66     12000
          weighted avg       0.79      0.80      0.80     12000

0.7985833333333333


In [51]:
#Accuracy of test and train
print('Accuracy on test data ', mlp.score(X_test,y_test))
print('Accuracy on train data ', mlp.score(X_train,y_train))

Accuracy on test data  0.7985833333333333
Accuracy on train data  0.9783958333333334


Report the accuracy, 
precision, 
recall, 
F1-score, 
confusion matrix.

# Accuracy

Accuracy on test data  0.7985833333333333
Accuracy on train data  0.9783958333333334


# Precision 

science_and_technology--0.55      
sports------------------0.66      
video_games-------------0.85

macro avg---------------0.69
weighted avg------------0.79

# Recall
science_and_technology--0.44
sports------------------0.61
video_games-------------0.88
macro avg---------------0.65
weighted avg------------0.80


# F1-score
science_and_technology--0.49
sports------------------0.64
video_games-------------0.87
macro avg---------------0.66
weighted avg------------0.80

# Confusion matrix
sciences_and_technology [[ 223   42  239]
sports_games             [  43 1781 1078]
video_games              [ 142  873 7579]]

# 2. Neural Network Classifier with Keras

Using the multi-label classifier dataset from earlier exercises (categorized-comments.jsonl in the reddit folder), fit a neural network classifier using Keras. Use the code found in chapter 12 of the Applied Text Analysis with Python book as a guideline. Report the accuracy, precision, recall, F1-score, and confusion matrix.

In [1]:
#Import libraries for keras tutorial
import keras
import tensorflow as tf
from numpy import loadtxt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import StandardScaler

In [39]:
#Read the file
jfile = "categorized-comments.jsonl" #File location
file = pd.read_json(jfile,lines=True) #Read command
file.head() #Check the datarame

cat                                                txt
0  sports  Barely better than Gabbert? He was significant...
1  sports  Fuck the ducks and the Angels! But welcome to ...
2  sports  Should have drafted more WRs.\n\n- Matt Millen...
3  sports            [Done](https://i.imgur.com/2YZ90pm.jpg)
4  sports                                      No!! NOO!!!!!

In [3]:
#Size of the dataframe
file.shape

(606476, 2)

In [4]:
#How are the categories split
file.cat.value_counts()

video_games               435542
sports                    145823
science_and_technology     25111
Name: cat, dtype: int64

In [40]:
#Break the single df into multiple df
video = file[file['cat'] == 'video_games']
sports = file[file['cat'] == 'sports']
sci_tech = file[file['cat'] == 'science_and_technology']

In [41]:
#Quick preview
print('Video file:',video)

print('Sports file:', sports)

print ('Sci and Tech file:', sci_tech)

Video file:                 cat                                                txt
65873   video_games                                           Nvm then
65874   video_games                             Hit speed 0,5 seconds.
65875   video_games  I don't have any. FFS THIS IS SO EASY TO SEE! ...
65876   video_games                                         On wheels?
65877   video_games         Nice!Good to see lumberjack in your deck 😀
...             ...                                                ...
606471  video_games             No. It's probably only happened to you
606472  video_games  I think most of the disappointment came from t...
606473  video_games  dishonored 1/2 looked like arse, so what the h...
606474  video_games                                          [removed]
606475  video_games  I wish more games provided options like Rise o...

[435542 rows x 2 columns]
Sports file:            cat                                                txt
0       sports  Barely better 

In [42]:
#Get samples 25000 sample from each
video_small = video.sample(25000)
sports_small = sports.sample(25000)
sci_tech_small = sci_tech.sample(25000)

In [43]:
#Put the files back into a single file with equal number of samples
# Concatenate data frames by categories into one
file_small_eql = pd.concat([video_small, sports_small, sci_tech_small])
file_small_eql.shape

(75000, 2)

In [44]:
#Check point copy
file_small_eql_copy1=file_small_eql.copy(deep=True)
#Check it again
file_small_eql.shape


(75000, 2)

In [13]:
#Use only if needed to backup a step
file_small_eql=file_small_eql_copy1.copy(deep=True)
#Check restored file
file_small_eql.shape

(75000, 2)

In [45]:
#Lowercase the data
file_small_eql["txt"] = file_small_eql["txt"].str.lower()
#View the top 5 rows

#Remove punctuation
file_small_eql["txt"] = file_small_eql["txt"].str.replace("[^\w\s]", "")
#View the top 5 rows
file_small_eql.head()

cat                                                txt
336645  video_games              i know this is days old but thank you
119394  video_games  yes an objective based game mode that forces t...
165323  video_games  httpsmsdnmicrosoftcomenuslibraryms176020aspxf2...
311732  video_games  thats what im saying when i said such a large ...
302492  video_games  i know its not quite on the same level but i n...

In [30]:
#Download stopwords and punkt
import nltk
nltk.download('stopwords')

import nltk
nltk.download('punkt')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Tushar\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Tushar\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [46]:
#Import NLKT, stopwords and tokenize
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 

#Set the stop_words
stop_words = stopwords.words("english") 

In [12]:
file_small_eql.shape

(75000, 2)

In [47]:
# Remove stop words
file_small_eql["txt"] = file_small_eql["txt"].apply(lambda x: " ".join([word for word in x.split()
                                                                        if word not in (stop_words)]))

In [14]:
file_small_eql.head

<bound method NDFrame.head of                            cat  \
389124             video_games   
402969             video_games   
568902             video_games   
385986             video_games   
335282             video_games   
...                        ...   
25824   science_and_technology   
7676    science_and_technology   
27345   science_and_technology   
17095   science_and_technology   
5518    science_and_technology   

                                                      txt  
389124                                         bloodborne  
402969                                   one day nintendo  
568902  yup struggle real blizzard puts lot faith play...  
385986  diablo game switch would amazing even torchlig...  
335282       sell goes back sold weeks ago 870k want back  
...                                                   ...  
25824   ill go ahead disagree priceperformance aspect ...  
7676    different would android experience s7e vs used...  
27345                

In [48]:
#Import PorterStemmer
from nltk.stem.porter import PorterStemmer

#Create stemmer
porter = PorterStemmer()

In [49]:
#Run stemmer
file_small_eql["txt"]= file_small_eql["txt"].apply(lambda x: " ".join([porter.stem(word) for word in x.split()]))
#View the file
file_small_eql.head()




cat                                                txt
336645  video_games                                 know day old thank
119394  video_games  ye object base game mode forc team move spawn win
165323  video_games  httpsmsdnmicrosoftcomenuslibraryms176020aspxf2...
311732  video_games  that im say said larg chunk busi wouldnt care ...
302492  video_games                    know quit level need sli cooper

In [50]:
# Use encoder to convert categories to numerical values
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

cat = file_small_eql["cat"]
file_small_eql["cat"]=encoder.fit_transform(cat)
#Check the values
file_small_eql.cat.value_counts()

2    25000
1    25000
0    25000
Name: cat, dtype: int64

In [51]:
#Bring in CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [52]:
#Create the bag of words features Features = X
count=CountVectorizer()
X=count.fit_transform(file_small_eql["txt"])
y = file_small_eql["cat"]
#Check it
X.shape

(75000, 44629)

In [53]:
#Target Preview
y=file_small_eql["cat"]
y.shape

(75000,)

In [54]:
#Quick preview
y

336645    2
119394    2
165323    2
311732    2
302492    2
         ..
8816      0
12951     0
19722     0
16215     0
12531     0
Name: cat, Length: 75000, dtype: int32

In [55]:
#Import train_test_split
from sklearn.model_selection import train_test_split

#Split the data into train and text dataset
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size = 0.20, random_state = 1)

In [56]:
#Preview the train test dataset
print('X Training set:',X_train.shape)

print ('Y Training set:',y_train.shape)

print ('X Test:',X_test.shape)

print ('Y Test:',y_test.shape)

X Training set: (60000, 44629)
Y Training set: (60000,)
X Test: (15000, 44629)
Y Test: (15000,)


In [57]:
X_train

<60000x44629 sparse matrix of type '<class 'numpy.int64'>'
	with 762848 stored elements in Compressed Sparse Row format>

In [58]:
# Convert to arrays for easy of use
y_train = np.asarray(y_train)
y_test = np.asarray(y_test)

In [59]:
#PReview
y_train.shape

(60000,)

In [60]:
#Function to convert to dense matrix
def tensor(X):
    coo = X.tocoo()
    indices = np.mat([coo.row, coo.col]).transpose()
    return tf.sparse.reorder(tf.SparseTensor(indices, coo.data, coo.shape))


# Define function to convert sparse matrix to sparse tensor
#def convert_sparse_matrix_to_sparse_tensor(X):
#    coo = X.tocoo()
#    indices = np.mat([coo.row, coo.col]).transpose()
#    return tf.sparse.reorder(tf.SparseTensor(indices, coo.data, coo.shape))

In [61]:
#Make them dense
X_train = tensor(X_train)
X_test = tensor(X_test)

In [62]:
#Check what comes out
X_train

In [63]:
#Initializing Neural Network
classifier = Sequential()

#Set input feature
N_FEATURES = X_train.shape[1]

In [64]:
#Create keras sequential classifier 

#Adding the input layer and the first hidden layer
classifier.add(Dense(500, activation='relu', input_dim=N_FEATURES))
#Adding the second hidden layer
classifier.add(Dense(700, activation='relu'))
#Adding the output layer
classifier.add(Dense(10, activation='softmax'))

#Compile the Artificial Neural Network
classifier.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])



In [65]:
#Fit training set
classifier.fit(X_train, y_train, batch_size=500, epochs=10, verbose=1)



Epoch 1/10
120/120 [==============================] - 13s 99ms/step - loss: 1.0372 - accuracy: 0.6224
Epoch 2/10
120/120 [==============================] - 15s 127ms/step - loss: 0.4031 - accuracy: 0.8257
Epoch 3/10
120/120 [==============================] - 16s 132ms/step - loss: 0.2985 - accuracy: 0.8656
Epoch 4/10
120/120 [==============================] - 16s 128ms/step - loss: 0.2343 - accuracy: 0.8938
Epoch 5/10
120/120 [==============================] - 15s 124ms/step - loss: 0.1793 - accuracy: 0.9176
Epoch 6/10
120/120 [==============================] - 16s 132ms/step - loss: 0.1455 - accuracy: 0.9323
Epoch 7/10
120/120 [==============================] - 17s 140ms/step - loss: 0.1270 - accuracy: 0.9376
Epoch 8/10
120/120 [==============================] - 16s 133ms/step - loss: 0.1149 - accuracy: 0.9449
Epoch 9/10
120/120 [==============================] - 16s 134ms/step - loss: 0.1081 - accuracy: 0.9471
Epoch 10/10
120/120 [==============================] - 17s 140ms/step - lo

In [66]:
#Create prediction
y_pred = np.argmax(classifier.predict(X_test), axis=-1)

In [67]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [68]:
#COnfusion Matrix
print(confusion_matrix(y_test, y_pred))

[[4021  440  558]
 [ 477 3706  733]
 [ 714  913 3438]]


In [69]:
#Classification Report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.77      0.80      0.79      5019
           1       0.73      0.75      0.74      4916
           2       0.73      0.68      0.70      5065

    accuracy                           0.74     15000
   macro avg       0.74      0.74      0.74     15000
weighted avg       0.74      0.74      0.74     15000



In [70]:
print(accuracy_score(y_test,y_pred))

0.7443333333333333


Accuray 74.43%

# 3. Classifying Images

In chapter 20 of the Machine Learning with Python Cookbook, implement the code found in section 20.15 classify MSINT images using a convolutional neural network. Report the accuracy of your results.

In [1]:
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers.convolutional import Conv2D, MaxPooling2D
from keras.utils import np_utils
from keras import backend as K

In [2]:
# Set that the color channel value will be first
K.set_image_data_format("channels_first")

In [3]:
# Set seed
np.random.seed(0)


In [4]:
# Set image information
channels = 1
height = 28
width = 28

In [5]:
# Load data and target from MNIST data
(data_train, target_train), (data_test, target_test) = mnist.load_data()

In [6]:
# Reshape training image data into features
data_train = data_train.reshape(data_train.shape[0], channels, height, width)

In [7]:
# Reshape test image data into features
data_test = data_test.reshape(data_test.shape[0], channels, height, width)

In [8]:
# Rescale pixel intensity to between 0 and 1
features_train = data_train / 255
features_test = data_test / 255

In [9]:
# One-hot encode target
target_train = np_utils.to_categorical(target_train)
target_test = np_utils.to_categorical(target_test)
number_of_classes = target_test.shape[1]

In [10]:
# Start neural network
network = Sequential()

In [11]:
# Add convolutional layer with 64 filters, a 5x5 window, and ReLU activation function
network.add(Conv2D(filters=64,
                   kernel_size=(5, 5),
                   input_shape=(channels, width, height),
                   activation='relu'))

In [12]:
# Add max pooling layer with a 2x2 window
network.add(MaxPooling2D(pool_size=(2, 2)))

In [13]:
# Add dropout layer
network.add(Dropout(0.5))

In [14]:
# Add layer to flatten input
network.add(Flatten())

In [15]:
# # Add fully connected layer of 128 units with a ReLU activation function
network.add(Dense(128, activation="relu"))

In [16]:
# Add dropout layer
network.add(Dropout(0.5))

In [17]:
# Add fully connected layer with a softmax activation function
network.add(Dense(number_of_classes, activation="softmax"))

In [18]:
# Compile neural network
network.compile(loss="categorical_crossentropy", # Cross-entropy
                optimizer="rmsprop", # Root Mean Square Propagation
                metrics=["accuracy"]) # Accuracy performance metric

In [40]:
# Train neural network
network.fit(features_train, # Features
            target_train, # Target
            epochs=2, # Number of epochs
            verbose=2, # Don't print description after each epoch
            batch_size=1000, # Number of observations per batch
            validation_data=(features_test, target_test)) # Data for evaluation

Epoch 1/2
60/60 - 40s - loss: 0.0822 - accuracy: 0.9758 - val_loss: 0.0441 - val_accuracy: 0.9860
Epoch 2/2
60/60 - 44s - loss: 0.0716 - accuracy: 0.9786 - val_loss: 0.0405 - val_accuracy: 0.9852


Report the accuracy of your results.
The accuracy 98.60 and 98.52 the two epochs ran.